nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/ScanKudu.java

```diff
@@ -0,0 +1,406 @@
+            // However, if we have no FlowFile and we have connections coming from other Processors, then
+            // we know that we should run only if we have a FlowFile.
+            if (fileToProcess == null && context.hasNonLoopConnection()) {
+                return;
+            }
+        }
+
+        projectedColumnNames = Arrays.asList(context.getProperty(PROJECTED_COLUMNS).evaluateAttributeExpressions(fileToProcess).getValue().split(","));
+        String predicate = context.getProperty(PREDICATES).evaluateAttributeExpressions(fileToProcess).getValue();
+        final String tableName = context.getProperty(TABLE_NAME).evaluateAttributeExpressions(fileToProcess).getValue();
+        if (fileToProcess == null) {
+            fileToProcess = session.create();
+        }
+        if (StringUtils.isBlank(tableName)) {
+            getLogger().error("Table Name is blank or null for {}, transferring to failure", new Object[] {fileToProcess});
+            session.transfer(fileToProcess, REL_FAILURE);
+            return;
+        }
+
+        try {
+            this.kuduTable = getKuduClient().openTable(tableName);
+        } catch (Exception e) {
+            getLogger().error("Unable to open Kudu table {} due to {}", new Object[] {tableName, e});
+            session.transfer(fileToProcess, REL_FAILURE);
+            return;
+        }
+
+        Integer batchSize = Integer.valueOf(context.getProperty(BATCH_SIZE).getValue());
+        final AtomicLong rowsPulledHolder = new AtomicLong(0);
+        final AtomicLong ffCountHolder = new AtomicLong(0);
+        ScanKuduResultHandler handler = new ScanKuduResultHandler(session, fileToProcess, rowsPulledHolder, ffCountHolder, tableName, batchSize);
+
+        try {
+            scan(context,
+                session,
+                this.kuduTable,
+                predicate,
+                projectedColumnNames,
+                handler);
+
+        } catch (Exception e) {
+            if (handler.getFlowFile() != null){
+                session.remove(handler.getFlowFile());
+            }
+            getLogger().error("Unable to fetch rows from Kudu table {} due to {}", new Object[] {tableName, e});
+            fileToProcess = session.putAttribute(fileToProcess, "scankudu.results.found", Boolean.toString(handler.isHandledAny()));
+            session.transfer(fileToProcess, REL_FAILURE);
+            return;
+        }
+
+        fileToProcess = session.putAttribute(fileToProcess, "scankudu.results.found", Boolean.toString(handler.isHandledAny()));
+
+        FlowFile openedFF = handler.getFlowFile();
+        if (openedFF != null) {
+            finalizeFlowFile(session, openedFF, tableName, handler.getRecordsCount(), null);
+        }
+
+        session.transfer(fileToProcess, REL_ORIGINAL);
+        session.commit();
+
+    }
+
+    /**
+     * Result Handler for Scan operation
+     */
+    private class ScanKuduResultHandler implements ResultHandler {
```

Comment:
After carefully looking into this it became apparent that the `AtomicLong` `rowsPulledHolder` and `ffCountHolder` properties are not handled in a thread-safe manner. They are not - and _cannot_ be - used concurrently.

Still the implementation is not incorrect if we take into account this constraint. However some simplification can be made as further checks reveal that the `ffCountHolder` property is not even used.
Also a unit test verifying the correctness of creating new flowfile when the batchsize limit has been reached would be definitely welcome.

The whole `ScanKuduResultHandler` could look like this:
```java
    private class ScanKuduResultHandler implements ResultHandler {

        final private ProcessSession session;
        final private FlowFile origFF;
        final private String tableName;
        final private Integer batchSize;
        private FlowFile flowFile = null;
        final private byte[] JSON_ARRAY_DELIM = ",\n".getBytes();

        private AtomicLong rowsPulledHolder = new AtomicLong(0);

        private boolean handledAny = false;

        ScanKuduResultHandler(final ProcessSession session,
                              final FlowFile origFF,
                              final String tableName, final Integer batchSize){
            this.session = session;
            this.tableName = tableName;
            this.batchSize = batchSize == null ? 0 : batchSize;
            this.origFF = origFF;

        }

        @Override
        public void handle(final RowResult resultCells) {

            long rowsPulled = rowsPulledHolder.get();

            try{
                if (flowFile == null){
                    flowFile = initNewFlowFile(session, origFF, tableName);
                }

                flowFile = session.append(flowFile, (out) -> {
                    if (rowsPulledHolder.get() > 0){
                        out.write(JSON_ARRAY_DELIM);
                    }
                    final String json = convertToJson(resultCells);
                    out.write(json.getBytes());
                });
                handledAny = true;

            } catch (Exception e) {
                throw new RuntimeException(e);
            }

            rowsPulled++;

            // bulkSize controls number of records per flow file.
            if (batchSize > 0 && rowsPulled >= batchSize) {

                finalizeFlowFile(session, flowFile, tableName, rowsPulled, null);
                flowFile = null;
                rowsPulledHolder.set(0);
                // we could potentially have a huge number of rows. If we get to batchSize, go ahead and commit the
                // session so that we can avoid buffering tons of FlowFiles without ever sending any out.
                session.commit();
            } else {
                rowsPulledHolder.set(rowsPulled);
            }
        }

        public boolean isHandledAny(){
            return handledAny;
        }

        @Override
        public FlowFile getFlowFile(){
            return flowFile;
        }

        public long getRecordsCount(){
            return rowsPulledHolder.get();
        }

    }
```